PythonVisualizations/OrderedArray.py

```diff
@@ -68,40 +68,75 @@ def createIndex(self, index,  # cell
     def insert(self, val):
         callEnviron = self.createCallEnvironment()
         self.startAnimations()
-        j = self.find(val)  # Find where item should go
-
+       # j = self.find(val)  # Find where item should go
+        k=len(self.list)
         self.list.append(drawable(None))
         
         indexK = self.createIndex(len(self.list) -1, 'k', level=-1) # create "k" arrow
         callEnviron |= set(indexK)  
         
-        for k in range(len(self.list) - 1, j, -1):  # Move bigger items right
-            
+       # for k in range(len(self.list) - 1, j, -1):  # Move bigger items right
+        while 0 < k and self.list[k-1].val > val: # over items
+
             self.wait(1)
             for item in indexK:
                 self.canvas.move(item, -self.CELL_SIZE, 0)  # Move "k" arrow  
-                
-            self.list[k].val = self.list[k - 1].val 
-            self.assignElement(k - 1, k, callEnviron)   
-            
+
+            self.list[k].val = self.list[k - 1].val # Move larger item to right
+            self.assignElement(k - 1, k, callEnviron)
+            k -= 1  # Advance left among items
         # Location of the new cell in the array
-        toPositions = (self.cellCoords(j),
-                       self.cellCenter(j))
+        toPositions = (self.cellCoords(k),
+                       self.cellCenter(k))
 
         # Animate arrival of new value from operations panel area
         canvasDimensions = self.widgetDimensions(self.canvas)
         startPosition = [canvasDimensions[0] // 2, canvasDimensions[1]] * 2
         startPosition = add_vector(startPosition, (0, 0, self.CELL_SIZE, self.CELL_SIZE))
         cellPair = self.createCellValue(startPosition, val)
         self.moveItemsTo(cellPair, toPositions, steps= self.CELL_SIZE, sleepTime=0.01)
-        self.list[j]= (drawable(
+        self.list[k]= (drawable(
             val, self.canvas.itemconfigure(cellPair[0], 'fill'), *cellPair))        
```

Comment:
The call to `self.canvas.itemconfigure(cellPair[0], 'fill')` returns a tuple in which the lat element is actual fill color.  To extract that color and put it in the `drawable` record, use `self.canvas.itemconfigure(cellPair[0], 'fill')[-1]`.

Separately, lines 95 and 96 introduce the new item to the canvas and then move it into position, but they need to mark it as temporary to this call until it replaces the item in the array.  This will prevent them from being left on the canvas if the user stops the animation before it ends.   Specifcally, those lines should be:
```
        cellPair = self.createCellValue(startPosition, val)
        callEnviron |= set(cellPair)           # Mark the new items as temporary
        self.moveItemsTo(cellPair, toPositions, steps= self.CELL_SIZE, sleepTime=0.01)
        self.canvas.destroy(self.list[k].display_shape)    # These are now covered by the temporary items
        self.canvas.destroy(self.list[k].display_val)
        self.list[k]= (drawable(               # Put temporary items into data structure
            val, self.canvas.itemconfigure(cellPair[0], 'fill'), *cellPair))   
        callEnviron ^= set(cellPair)           # New item is no longer temporary 
```